### PR TITLE
Add script to migrate feed util to cron keeper

### PR DESCRIPTION
**Pull request by infiloop2**
This script will be used to migrate feed util jobs to cron upkeeps

- It takes in target address, target function, cron schedule and additional registration parameters as input
- deploys a new cron upkeep contract using cron factory
- sets permission on authorised forwarder for this new contract to call target
- requests upkeep registration on registrar
- stores the cronUpkeep contract address and registration hash to output file

- In a separate command, it takes in the previous output file as input and fetches the upkeep ID if it has been approved

-----------------------------------
Notes:
- Following env config is needed in .env file for the script to work
    - CHAIN_ID
    - NODE_URL
    - PRIVATE_KEY
    - LINK_TOKEN_ADDR
    - CRON_FACTORY_ADDR
    - REGISTRAR_ADDR
    - FORWARDER_ADDR
- The account associated with private key should hold sufficient LINK and also be the owner of Forwarder contract
- More info on the script can be found by ``` go run main.go keeper migrate-cron --help```. It specifies the format of input and output files
- Script process input line by line and outputs along the way. In case it crashes in between, the processed rows should have an associated output. In the next run, those rows can be removed from input.
- Setting the permission on forwarder requires us to read the existing list, append a new address and then update the full new list. This means that addresses will be overwritten if script is run in parallel, take caution.
-----------------------------------

Example usage of script using sample input file
```
x86 | chaincli $ go run main.go keeper migrate-cron --input-file migrate_cron_input.csv.example
2022/04/11 19:44:22 Version was unset but CHAINLINK_DEV was not set to "true". Chainlink should be built with static.Version set to a valid semver for production builds.
2022/04/11 19:44:22 Using config file .env
Processing: 0x32f8F3021F36558f7822A283cddD3C7C1Eae9071 handler1 * * * * *
2022/04/11 19:44:22 Deploying new cron upkeep
2022/04/11 19:44:33 Cron upkeep deployed: https://kovan.etherscan.io/tx/0x98ed4c7125d0797f986fc954007c2f79cf11c2d35813d939f6e96b816187ab8b
2022/04/11 19:44:33 Cron upkeep address: 0xC2A81dE1b6Ddc74347041421a456130c7db0900E
2022/04/11 19:44:33 Setting permission on proxy
2022/04/11 19:44:41 Proxy permission from 0xC2A81dE1b6Ddc74347041421a456130c7db0900E to 0x32f8F3021F36558f7822A283cddD3C7C1Eae9071 set: https://kovan.etherscan.io/tx/0x7e7cfb860e4ecc479577f56277202dd1d41c95c606b2498cd88e132481224d06
2022/04/11 19:44:41 Registering upkeep
2022/04/11 19:44:49 Upkeep registered: https://kovan.etherscan.io/tx/0x3d5941563f8f7d9c93370d6b87eb90c847cf7058342bfca77486f284471ded34
2022/04/11 19:44:49 Registration request hash: 52fb17ff09b75febc8393e5f91ca9a5a1f49b43d5b3d7fd23b26701f2fa3f6f2
Processing: 0x32f8F3021F36558f7822A283cddD3C7C1Eae9071 handler1 * * * * *
2022/04/11 19:44:49 Deploying new cron upkeep
2022/04/11 19:44:57 Cron upkeep deployed: https://kovan.etherscan.io/tx/0xca572d6eff3ca8ac4206f52484c99f606eed0ee6cdc6d570f466fee8ce6aa068
2022/04/11 19:44:57 Cron upkeep address: 0x0AEdC107695EE5768e4178ADA9929dC5CA0eD84A
2022/04/11 19:44:57 Setting permission on proxy
2022/04/11 19:45:05 Proxy permission from 0x0AEdC107695EE5768e4178ADA9929dC5CA0eD84A to 0x32f8F3021F36558f7822A283cddD3C7C1Eae9071 set: https://kovan.etherscan.io/tx/0xea3498b31946654c3d449f9c9cfa1b1310252be0c5398a50fcbaa02cda86b2c6
2022/04/11 19:45:05 Registering upkeep
2022/04/11 19:45:12 Upkeep registered: https://kovan.etherscan.io/tx/0xf1c26a86fe2bd3040f367024bffa4a9f30b7aaec7930eff31b7574df94932b46
2022/04/11 19:45:12 Registration request hash: b075146eab11fed12df2daac89fe1b0ff98aaff31bf23a9e71d9b2e8df091596
```

Approved the upkeeps and fetched the upkeep IDs with '-f' flag

```
x86 | chaincli $ go run main.go keeper migrate-cron --input-file migrate_cron_output_2022-04-11\ 19:34:40.16129\ +0100\ BST\ m=+0.088385084.csv -f
2022/04/11 19:41:07 Version was unset but CHAINLINK_DEV was not set to "true". Chainlink should be built with static.Version set to a valid semver for production builds.
2022/04/11 19:41:07 Using config file .env
2022/04/11 19:41:09 Found upkeep ID for hash 9741323cc67c86f31aee655451ff134aec6a562ad996d321042326a1418cbf4e 20
2022/04/11 19:41:09 Found upkeep ID for hash 6f7122d38eb645089eb2446022e65a1ad17c1cddce12b107a7ca77444fbd86da 21
```

### CODE REVIEW

File: core/scripts/chaincli/handler/migrate_cron.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+)
+
+func (k *Keeper) MigrateCron(ctx context.Context, inputFilePath string) {
+	proxyAddr := common.HexToAddress(k.cfg.ProxyAddr)
+	proxyAbi, err := abi.JSON(strings.NewReader(proxy.PermissionedForwardProxyABI))
+	if err != nil {
+		log.Fatalln("Error generating proxy ABI", err)
+	}
+
+	// Throw error if output file already exists
```
Comment (cmalec) [inline core/scripts/chaincli/handler/migrate_cron.go:30]: could just throw a timestamp in the file name if you wanted to allow the command to run twice without having the user delete the file after.

**Comment by RyanRHall** [comment]
Is this something we expect to use and maintain long-term? Or is it more like we use the script to migrate all of the feeds and then we can don't need it anymore? Just wondering if this should live long term on the develop branch or if we can just leave it in a feature branch until we're done with it.

File: core/scripts/chaincli/handler/migrate_cron.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	outputFile, err := os.OpenFile("migrate_cron_output.csv", os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666)
```
Comment (cmalec) [inline core/scripts/chaincli/handler/migrate_cron.go:31]: are we missing an explicit close for this?

File: core/scripts/chaincli/handler/migrate_cron.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+	"log"
+	"math/big"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	cron_factory "github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/cron_upkeep_factory_wrapper"
```
Comment (begmaroman) [inline core/scripts/chaincli/handler/migrate_cron.go:17]: Cosmetic: add empty line above

File: core/scripts/chaincli/handler/migrate_cron.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	proxy "github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/permissioned_forward_proxy_wrapper"
+	registrar "github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/upkeep_registration_requests_wrapper"
+	helpers "github.com/smartcontractkit/chainlink/core/scripts/common"
+)
+
+func (k *Keeper) MigrateCron(ctx context.Context, inputFilePath string) {
+	proxyAddr := common.HexToAddress(k.cfg.ProxyAddr)
+	proxyAbi, err := abi.JSON(strings.NewReader(proxy.PermissionedForwardProxyABI))
```
Comment (begmaroman) [inline core/scripts/chaincli/handler/migrate_cron.go:25]: Does it make sense to define `proxyAbi ` in the init function?

File: core/scripts/chaincli/handler/migrate_cron.go (hunk 1 above)
Comment (infiloop2) [reply core/scripts/chaincli/handler/migrate_cron.go:30]: good idea!

File: core/scripts/chaincli/handler/migrate_cron.go (hunk 2 above)
Comment (infiloop2) [reply core/scripts/chaincli/handler/migrate_cron.go:31]: (y) yep, i think i missed it

File: core/scripts/chaincli/handler/migrate_cron.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	cronFactoryInstance, err := cron_factory.NewCronUpkeepFactory(
+		cronFactoryAddr,
+		k.client,
+	)
+	if err != nil {
+		log.Fatalln("Error while instantiating "+cronFactoryAddr.String()+" to cron factory", err)
+	}
+
+	callOpts := bind.CallOpts{
+		Pending: false,
```
Comment (begmaroman) [inline core/scripts/chaincli/handler/migrate_cron.go:85]: You can just remove it, it is false by default

**Comment by infiloop2** [comment]
@RyanRHall I think you're right, we don't need to maintain this in develop. Just use it once to migrate the existing ones and new ones should be created using cron UI.

File: core/scripts/chaincli/handler/migrate_cron.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+	records, err := csvReader.ReadAll()
+	if err != nil {
+		log.Fatal("Unable to parse file as CSV for "+filePath, err)
+	}
+
+	return records
+}
+
+func getTargetHandler(targetFunction string) []byte {
+	targetAbi, err := abi.JSON(strings.NewReader(`[{"type":"function","inputs":[],"outputs":[],"name":"` + targetFunction + `"}]`))
```
Comment (RyanRHall) [inline core/scripts/chaincli/handler/migrate_cron.go:219]: is `targetFunction` sensitive to things like parens? ex will `handler1` and `handler1()` produce the same results?

Comment (infiloop2) [reply core/scripts/chaincli/handler/migrate_cron.go:219]: Good question, it works without (), but not sure what happens with them

File: core/scripts/chaincli/config/config.go [hunk 7]
```diff
@@ -8,19 +8,20 @@ import (
...
-	ChainID         int64    `mapstructure:"CHAIN_ID"`
-	PrivateKey      string   `mapstructure:"PRIVATE_KEY"`
-	LinkTokenAddr   string   `mapstructure:"LINK_TOKEN_ADDR"`
+	// Global Config
+	NodeURL       string `mapstructure:"NODE_URL"`
+	ChainID       int64  `mapstructure:"CHAIN_ID"`
+	PrivateKey    string `mapstructure:"PRIVATE_KEY"`
+	LinkTokenAddr string `mapstructure:"LINK_TOKEN_ADDR"`
+	GasLimit      uint64 `mapstructure:"GAS_LIMIT"`
+
```
[the remark this answers is not in the document]
Comment (infiloop2) [reply core/scripts/chaincli/config/config.go:17]: I'll probably pull this out in a separate PR and merge to develop

File: core/scripts/chaincli/handler/migrate_cron.go (hunk 6 above)
Comment (infiloop2) [reply core/scripts/chaincli/handler/migrate_cron.go:219]: It needs to be strictly the function name without parenthesis. Added an input validation for that

**Pull request closed without merging**